### PR TITLE
feat: merge package `credentials` back from `oras-credentials-go`

### PR DESCRIPTION
**Pull request by Wwwsylvia**
1. Add package `credentials` (migrated from [`oras-credentials-go`](https://github.com/oras-project/oras-credentials-go)) and its corresponding description
2. Add new type `auth.CredentialFunc`
3. Migrate the credentials example from [oras.land](https://oras.land/docs/client_libraries/go/#pull-an-image-using-the-docker-credential-store)
4. Change `auth.DefaultCache` to `auth.NewCache()` in examples

Resolves: #584

### CODE REVIEW

File: registry/remote/auth/client.go [hunk 1]
```diff
@@ -54,8 +54,15 @@ var maxResponseBytes int64 = 128 * 1024 // 128 KiB
 // See also ClientID.
 var defaultClientID = "oras-go"
 
+// CredentialFunc represents a function that resolves the credential for the
+// given registry (i.e. host:port).
+//
+// `EmptyCredential` is a valid return value and should not be considered as
+// an error.
+type CredentialFunc func(ctx context.Context, registry string) (Credential, error)
```
Comment (ridhoq) [inline registry/remote/auth/client.go:62]: Given the comment on the type, it might make more sense to name it `CredentialResolver`

Comment (Wwwsylvia) [reply registry/remote/auth/client.go:62]: Yeah, but I feel that `CredentialFunc` matches more with the name `StaticCredential`
https://github.com/oras-project/oras-go/blob/deb17e4896db7b46f6e8a290ae64a74ae20c4894/registry/remote/auth/client.go#L65

and `credentials.Credential`.
https://github.com/oras-project/oras-go/blob/deb17e4896db7b46f6e8a290ae64a74ae20c4894/registry/remote/credentials/registry.go#L72

It's more concise and straightforward. 🤔

File: registry/remote/credentials/trace/trace.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trace
+
+import (
+	"context"
+)
```
Comment (shizhMSFT) [inline registry/remote/credentials/trace/trace.go:20]: nit:
```suggestion
import "context"
```

File: registry/remote/auth/client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Wwwsylvia) [reply registry/remote/auth/client.go]: Let's call it `target` then.

File: registry/remote/auth/client.go [hunk 3]
```diff
@@ -57,9 +57,9 @@ var defaultClientID = "oras-go"
 // CredentialFunc represents a function that resolves the credential for the
 // given registry (i.e. host:port).
 //
-// `EmptyCredential` is a valid return value and should not be considered as
+// [EmptyCredential] is a valid return value and should not be considered as
 // an error.
-type CredentialFunc func(ctx context.Context, registry string) (Credential, error)
+type CredentialFunc func(ctx context.Context, target string) (Credential, error)
```
Comment (shizhMSFT) [inline registry/remote/auth/client.go:62]: `target` is ambiguous. It is indeed `hostname` and `port`.

How about `hostport` as [net.SplitHostPort()](https://pkg.go.dev/net#SplitHostPort)?

Comment (Wwwsylvia) [reply registry/remote/auth/client.go]: That works.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**